Fix installation issues in setup.sh

bearstonem opened the pull request:
## Summary
This PR fixes several critical installation issues in the setup.sh script that prevent successful installation of TRELLIS.2 dependencies.

## Issues Fixed

### 1. CUDA Linker Errors
**Problem:** Build fails with `ld: cannot find -lcuda: No such file or directory` when compiling nvdiffrec and cumesh
**Solution:** 
- Added proper LIBRARY_PATH and LDFLAGS configuration to include system library paths
- Includes both `/usr/lib/x86_64-linux-gnu` and CUDA stubs directory

### 2. Git Submodule SSH Authentication Failures
**Problem:** CuMesh submodule clone fails with `Permission denied (publickey)` error
**Solution:**
- Automatically convert git submodule URLs from SSH to HTTPS
- Sync and initialize submodules after URL conversion

### 3. CUDA_HOME Not Set
**Problem:** Script assumes CUDA_HOME is set, causing issues when it's not
**Solution:**
- Added auto-detection logic with fallback to common CUDA paths
- Tries `/usr/local/cuda-12.4` first, then `/usr/local/cuda`
- Provides helpful warning messages if detection fails

### 4. o-voxel Dependency Conflicts
**Problem:** o-voxel tries to reinstall cumesh and flex_gemm from git, causing failures
**Solution:**
- Modified pyproject.toml to use already-installed packages
- Prevents redundant git installations

## Changes Made

### setup.sh
- Added CUDA_HOME auto-detection (lines 69-87)
- Fixed nvdiffrec installation with proper library paths (lines 135-137)
- Fixed CuMesh submodule SSH issues (lines 156-160)
- Fixed o-voxel dependency conflicts (lines 178-181)
- Added informative logging throughout installation process

### README.md
- Added documentation about CUDA path auto-configuration
- Added troubleshooting notes for common linker issues
- Documented SSH to HTTPS conversion behavior

## Testing
Tested on Ubuntu with:
- CUDA 12.4 and 12.9 installations
- Multiple CUDA toolkit versions present
- Fresh installation without CUDA_HOME set

All packages now install successfully without manual intervention.

## Related Issues
This addresses installation failures that users commonly encounter when following the setup instructions, particularly on systems with:
- Multiple CUDA versions installed
- No SSH keys configured for GitHub
- Standard Linux library paths for NVIDIA drivers

Commit:
Fix installation issues in setup.sh

- Add CUDA_HOME auto-detection with fallback to common paths
- Fix linker errors by setting LIBRARY_PATH and LDFLAGS for libcuda.so
- Convert git submodule URLs from SSH to HTTPS to avoid auth issues
- Fix o-voxel dependencies to prevent reinstalling cumesh/flex_gemm
- Add informative logging messages during installation
- Update README with troubleshooting notes

Fixes:
- 'ld: cannot find -lcuda' linker errors during nvdiffrec/cumesh build
- 'Permission denied (publickey)' errors when cloning CuMesh submodules
- Dependency conflicts when installing o-voxel

Co-Authored-By: Warp <agent@warp.dev>

alexmolskiy commented:
Need to add `--no-build-isolation` to line 89 `pip install flash-attn==2.7.3`

bearstonem commented:
> @bearstonem please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.



> @bearstonem please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

Commit:
Add --no-build-isolation flag to flash-attn installation

This ensures flash-attn builds correctly with the existing environment.

ben-mkiv commented:
FYI, gentoo seems to put cuda to `/opt/cuda` which is likely also a very reasonable path if people install it manually, so maybe you can add that

bearstonem reviewed (commented):
looks good

Commit:
Fix o-voxel pyproject.toml sed: don't double-quote replacement

The dependency entries in o-voxel/pyproject.toml live inside a quoted TOML
array, like:

    dependencies = [
        "cumesh @ git+https://github.com/JeffreyXiang/CuMesh.git",
        ...
    ]

The previous sed replaced `cumesh @ git+...` with `"cumesh"` (literal quotes
in the replacement), producing `""cumesh""` — invalid TOML. pip then fails
with `tomllib.TOMLDecodeError: Unclosed array`.

Strip the quotes from the replacement so the outer pair is preserved.

Commit:
flash-attn: install prebuilt 2.8.3 wheel instead of source-building 2.7.3

flash-attn 2.7.3 has no prebuilt wheel for torch 2.7 (release ships
2.1-2.6 + 2.8 only), and source-building against torch 2.7.1+cu128 fails
under modern setuptools/wheel during egg-info collection.

2.8.3 publishes an exact `cu12torch2.7cxx11abiFALSE` wheel for cp311
(and other cp3x). The bumped version is API-compatible for the dense
and varlen attention paths TRELLIS.2 uses.
